File: frost-taproot-go/sign/sign.go
// Package sign provides signing and signature verification.
package sign

import (
	"bytes"
	"math/big"

	"github.com/frost-taproot/frost-taproot-go/commit"
	"github.com/frost-taproot/frost-taproot-go/ecc"
	"github.com/frost-taproot/frost-taproot-go/helpers"
	"github.com/frost-taproot/frost-taproot-go/poly"
	"github.com/frost-taproot/frost-taproot-go/types"
	"github.com/frost-taproot/frost-taproot-go/util"
)

// SignMsg produces a partial signature for a signing session.
func SignMsg(ctx *types.GroupSigningCtx, share *types.SecretShare, snonce *types.SecretNonce) (types.ShareSignature, error) {
	if snonce.ID != share.ID {
		return types.ShareSignature{}, &util.AssertionError{Message: "commit index does not match share index"}
	}

	bindFactorBytes, err := commit.GetBindFactor(ctx.BindFactors, share.ID)
	if err != nil {
		return types.ShareSignature{}, err
	}
	bindFactor := ecc.ScalarFromBytes(bindFactorBytes)

	indexes := make([]*big.Int, len(ctx.Indexes))
	for i, idx := range ctx.Indexes {
		indexes[i] = poly.IndexToScalar(idx)
	}
	coefficient, err := poly.InterpolateX(indexes, poly.IndexToScalar(share.ID))
	if err != nil {
		return types.ShareSignature{}, err
	}

	snonceH := ecc.ScalarFromBytes(snonce.HiddenSn)
	snonceB := ecc.ScalarFromBytes(snonce.BinderSn)
	seckey := ecc.ScalarFromBytes(share.Seckey)

	rElem, err := ecc.LiftX(ctx.GroupPn[:])
	if err != nil {
		return types.ShareSignature{}, err
	}

	if !ecc.HasEvenY(rElem) {
		snonceH = ecc.ScalarNeg(snonceH)
		snonceB = ecc.ScalarNeg(snonceB)
	}

	// sk = parity * state * seckey
	parity := ecc.ScalarFromBytes(ctx.GroupPt.Parity)
	state := ecc.ScalarFromBytes(ctx.GroupPt.State)
	sk := ecc.ScalarMul(ecc.ScalarMul(parity, state), seckey)

	// nk = hidden_sn + binder_sn * bind_factor
	nk := ecc.ScalarAdd(snonceH, ecc.ScalarMul(snonceB, bindFactor))

	// ps = challenge * coefficient * sk + nk
	challenge := ecc.ScalarFromBytes(ctx.Challenge)
	ps := ecc.ScalarAdd(ecc.ScalarMul(ecc.ScalarMul(challenge, coefficient), sk), nk)

	return types.ShareSignature{
		ID:     share.ID,
		Psig:   ecc.ScalarToBytes(ps),
		Pubkey: helpers.GetPubkey(share.Seckey),
	}, nil
}

// CombinePartialSigs aggregates partial signatures into a final BIP340 Schnorr signature.
func CombinePartialSigs(ctx *types.GroupSigningCtx, psigs []types.ShareSignature) ([64]byte, error) {
	commitPrefix := commit.GetGroupPrefix(ctx.Pnonces, ctx.GroupPk, ctx.Message)
	groupBinders := commit.GetGroupBinders(ctx.Pnonces, commitPrefix)
	groupPnonce, err := commit.GetGroupPubnonce(ctx.Pnonces, groupBinders)
	if err != nil {
		return [64]byte{}, err
	}

	// Sum all partial signatures
	ps := big.NewInt(0)
	for _, s := range psigs {
		ps = ecc.ScalarAdd(ps, ecc.ScalarFromBytes(s.Psig))
	}

	// twk = challenge * parity * tweak
	challenge := ecc.ScalarFromBytes(ctx.Challenge)
	parity := ecc.ScalarFromBytes(ctx.GroupPt.Parity)
	tweak := ecc.ScalarFromBytes(ctx.GroupPt.Tweak)
	twk := ecc.ScalarMul(ecc.ScalarMul(challenge, parity), tweak)
	s := ecc.ScalarAdd(ps, twk)

	// Signature = R_x (32 bytes) || s (32 bytes)
	var sig [64]byte
	copy(sig[:32], groupPnonce[1:])
	sBytes := ecc.ScalarToBytes(s)
	copy(sig[32:], sBytes[:])
	return sig, nil
}

// VerifyPartialSig verifies a partial signature from one participant.
func VerifyPartialSig(ctx *types.GroupSigningCtx, pnonce *types.PublicNonce, sharePk [33]byte, sharePsig [32]byte) (bool, error) {
	binderBytes, err := commit.GetBindFactor(ctx.BindFactors, pnonce.ID)
	if err != nil {
		return false, err
	}
	binder := ecc.ScalarFromBytes(binderBytes)

	hiddenElem, err := ecc.LiftX(pnonce.HiddenPn[:])
	if err != nil {
		return false, err
	}
	binderElem, err := ecc.LiftX(pnonce.BinderPn[:])
	if err != nil {
		return false, err
	}
	publicElem, err := ecc.LiftX(sharePk[:])
	if err != nil {
		return false, err
	}

	rElem, err := ecc.LiftX(ctx.GroupPn[:])
	if err != nil {
		return false, err
	}

	if !ecc.HasEvenY(rElem) {
		hiddenElem = ecc.NegatePoint(hiddenElem)
		binderElem = ecc.NegatePoint(binderElem)
	}

	commitElem := ecc.ScalarMulti(binderElem, binder)
	nonceElem, err := ecc.ElementAdd(hiddenElem, commitElem)
	if err != nil {
		return false, err
	}

	indexes := make([]*big.Int, len(ctx.Indexes))
	for i, idx := range ctx.Indexes {
		indexes[i] = poly.IndexToScalar(idx)
	}
	lambdaI, err := poly.InterpolateX(indexes, poly.IndexToScalar(pnonce.ID))
	if err != nil {
		return false, err
	}

	parity := ecc.ScalarFromBytes(ctx.GroupPt.Parity)
	state := ecc.ScalarFromBytes(ctx.GroupPt.State)
	chal := ecc.ScalarMul(ecc.ScalarMul(ecc.ScalarFromBytes(ctx.Challenge), lambdaI), ecc.ScalarMul(parity, state))

	sig := ecc.ScalarFromBytes(sharePsig)
	sg := ecc.ScalarBaseMulti(sig)
	pki := ecc.ScalarMulti(publicElem, chal)
	r, err := ecc.ElementAdd(nonceElem, pki)
	if err != nil {
		return false, err
	}

	return bytes.Equal(sg.X.Bytes(), r.X.Bytes()), nil
}

// VerifyFinalSig verifies a final aggregated BIP340 Schnorr signature.
func VerifyFinalSig(ctx *types.GroupKeyContext, message []byte, signature [64]byte) (bool, error) {
	// BIP340 verification
	// group_pk is 33-byte compressed; BIP340 uses x-only (32 bytes)
	pkBytes := ctx.GroupPk[1:]

	// Parse R from signature (first 32 bytes)
	rX := new(big.Int).SetBytes(signature[:32])
	rY := new(big.Int).ModSqrt(
		new(big.Int).Add(
			new(big.Int).Exp(rX, big.NewInt(3), ecc.P),
			big.NewInt(7),
		),
		ecc.P,
	)
	if rY == nil {
		return false, nil
	}
	if rY.Bit(0) == 1 {
		rY = new(big.Int).Sub(ecc.P, rY)
	}

	// Parse s from signature (last 32 bytes)
	s := new(big.Int).SetBytes(signature[32:])

	// Parse public key
	pkX := new(big.Int).SetBytes(pkBytes)
	pkY := new(big.Int).ModSqrt(
		new(big.Int).Add(
			new(big.Int).Exp(pkX, big.NewInt(3), ecc.P),
			big.NewInt(7),
		),
		ecc.P,
	)
	if pkY == nil {
		return false, nil
	}
	if pkY.Bit(0) == 1 {
		pkY = new(big.Int).Sub(ecc.P, pkY)
	}

	// Compute challenge
	challenge, err := helpers.GetChallenge(
		signature[:32],
		ctx.GroupPk[:],
		message,
	)
	if err != nil {
		return false, err
	}

	// Verify: s*G = R + challenge*P
	sG := ecc.ScalarBaseMulti(s)
	cP := ecc.ScalarMulti(&ecc.Point{X: pkX, Y: pkY}, challenge)
	rPoint := &ecc.Point{X: rX, Y: rY}
	rhs, err := ecc.ElementAdd(rPoint, cP)
	if err != nil {
		return false, err
	}

	return sG.X.Cmp(rhs.X) == 0 && sG.Y.Cmp(rhs.Y) == 0, nil
}

File: frost-taproot-rust/src/sign.rs
// Mirrors ref/frost/src/lib/sign.ts

use k256::{Scalar, elliptic_curve::point::AffineCoordinates};

use crate::Error;
use crate::commit::{get_bind_factor, get_group_binders, get_group_prefix, get_group_pubnonce};
use crate::ecc::group::{element_add, scalar_base_multi, scalar_multi};
use crate::ecc::util::{has_even_y, lift_x, scalar_from_bytes, scalar_to_bytes};
use crate::helpers::get_pubkey;
use crate::poly::{index_to_scalar, interpolate_x};
use crate::types::{
    GroupKeyContext, GroupSigningCtx, PublicNonce, SecretNonce, SecretShare, ShareSignature,
};

/// Produce a partial signature for a signing session.
/// Mirrors `sign_msg` in the TS implementation.
pub fn sign_msg(
    ctx: &GroupSigningCtx,
    share: &SecretShare,
    snonce: &SecretNonce,
) -> Result<ShareSignature, Error> {
    if snonce.idx != share.idx {
        return Err(Error::Assertion(format!(
            "commit index does not match share index: {} !== {}",
            snonce.idx, share.idx
        )));
    }

    let bind_factor_bytes = get_bind_factor(&ctx.bind_factors, share.idx)?;
    let bind_factor = scalar_from_bytes(&bind_factor_bytes);

    let indexes: Vec<Scalar> = ctx.indexes.iter().map(|&i| index_to_scalar(i)).collect();
    let coefficient = interpolate_x(&indexes, index_to_scalar(share.idx))?;

    let mut snonce_h = scalar_from_bytes(&snonce.hidden_sn);
    let mut snonce_b = scalar_from_bytes(&snonce.binder_sn);
    let seckey = scalar_from_bytes(&share.seckey);

    let r_elem = lift_x(&ctx.group_pn)?;
    if !has_even_y(&r_elem) {
        snonce_h = -snonce_h;
        snonce_b = -snonce_b;
    }

    // sk = parity * state * seckey  (mod N)
    let sk = ctx.group_pt.parity * ctx.group_pt.state * seckey;
    // nk = hidden_sn + binder_sn * bind_factor
    let nk = snonce_h + snonce_b * bind_factor;
    // ps = challenge * coefficient * sk + nk
    let ps = ctx.challenge * coefficient * sk + nk;

    Ok(ShareSignature {
        idx: share.idx,
        psig: scalar_to_bytes(&ps),
        pubkey: get_pubkey(&share.seckey),
    })
}

/// Aggregate partial signatures into a final BIP340 Schnorr signature (64 bytes).
/// Mirrors `combine_partial_sigs` in the TS implementation.
pub fn combine_partial_sigs(
    ctx: &GroupSigningCtx,
    psigs: &[ShareSignature],
) -> Result<[u8; 64], Error> {
    let commit_prefix = get_group_prefix(&ctx.pnonces, &ctx.group_pk, &ctx.message);
    let group_binders = get_group_binders(&ctx.pnonces, &commit_prefix);
    let group_pnonce = get_group_pubnonce(&ctx.pnonces, &group_binders)?;

    // Sum all partial signatures.
    let ps = psigs
        .iter()
        .map(|s| scalar_from_bytes(&s.psig))
        .fold(Scalar::ZERO, |acc, s| acc + s);

    // twk = challenge * parity * tweak
    let twk = ctx.challenge * ctx.group_pt.parity * ctx.group_pt.tweak;
    let s = ps + twk;

    // Signature = R_x (32 bytes) || s (32 bytes).
    // group_pnonce is a 33-byte compressed point; x-only is bytes [1..33].
    let mut sig = [0u8; 64];
    sig[..32].copy_from_slice(&group_pnonce[1..]);
    sig[32..].copy_from_slice(&scalar_to_bytes(&s));
    Ok(sig)
}

/// Verify a partial signature from one participant.
/// Mirrors `verify_partial_sig` in the TS implementation.
pub fn verify_partial_sig(
    ctx: &GroupSigningCtx,
    pnonce: &PublicNonce,
    share_pk: &[u8; 33],
    share_psig: &[u8; 32],
) -> Result<bool, Error> {
    let binder = scalar_from_bytes(&get_bind_factor(&ctx.bind_factors, pnonce.idx)?);

    let mut hidden_elem = lift_x(&pnonce.hidden_pn)?;
    let mut binder_elem = lift_x(&pnonce.binder_pn)?;
    let public_elem = lift_x(share_pk)?;

    let r_elem = lift_x(&ctx.group_pn)?;
    if !has_even_y(&r_elem) {
        hidden_elem = -hidden_elem;
        binder_elem = -binder_elem;
    }

    let commit_elem = scalar_multi(&binder_elem, &binder);
    let nonce_elem = element_add(Some(hidden_elem), Some(commit_elem))?;

    let indexes: Vec<Scalar> = ctx.indexes.iter().map(|&i| index_to_scalar(i)).collect();
    let lambda_i = interpolate_x(&indexes, index_to_scalar(pnonce.idx))?;

    let state = ctx.group_pt.parity * ctx.group_pt.state;
    let chal = ctx.challenge * lambda_i * state;

    let sig = scalar_from_bytes(share_psig);
    let sg = scalar_base_multi(&sig);
    let pki = scalar_multi(&public_elem, &chal);
    let r = element_add(Some(nonce_elem), Some(pki))?;

    // Compare x-coordinates.
    Ok(sg.to_affine().x() == r.to_affine().x())
}

/// Verify a final aggregated BIP340 Schnorr signature.
/// Mirrors `verify_final_sig` in the TS implementation.
///
/// Uses `verify_raw` because noble's `schnorr.verify` passes the message directly
/// into the BIP340 challenge hash without pre-hashing it with SHA-256.
pub fn verify_final_sig(
    ctx: &GroupKeyContext,
    message: &[u8],
    signature: &[u8; 64],
) -> Result<bool, Error> {
    use k256::schnorr::{Signature, VerifyingKey};

    // group_pk is 33-byte compressed; BIP340 uses x-only (32 bytes).
    let pk_bytes: [u8; 32] = ctx.group_pk[1..].try_into().unwrap();
    let vk = VerifyingKey::from_bytes(&pk_bytes).map_err(|_| Error::InvalidPoint)?;
    let sig = Signature::try_from(signature.as_slice()).map_err(|_| Error::InvalidPoint)?;

    Ok(vk.verify_raw(message, &sig).is_ok())
}
